eventually-v0-5-0/src/event.rs

```diff
@@ -0,0 +1,111 @@
+//! Module `event` contains types and abstractions helpful for working
+//! with Domain Events.
+
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use futures::stream::BoxStream;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    version::{ToConflictError, Version},
+    Message, Messages,
+};
+
+/// An Event is a [Message] carring the information about a Domain Event,
+/// an occurrence in the system lifetime that is relevant for the Domain
+/// that is being implemented.
+pub type Event<T> = Message<T>;
+
+/// Shortcut type to represent multiple [Event]s.
+pub type Events<T> = Messages<T>;
+
+/// An [Event] that has been persisted to the Event [Store].
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Persisted<Id, Evt> {
+    /// The id of the Event Stream the persisted Event belongs to.
+    pub stream_id: Id,
+
+    /// The version of the Event Stream when this Event has been recorded.
+    ///
+    /// This value is used for optimistic concurrency checks, to avoid
+    /// data races in parallel command evaluations.
+    ///
+    /// Check the [Version] type and module documentation for more info.
+    pub version: Version,
+
+    /// The actual Domain Event carried by this envelope.
+    pub payload: Event<Evt>,
+}
+
+/// Shortcut type to represent multiple [Persisted] Events.
+pub type PersistedEvents<Id, Evt> = Vec<Persisted<Id, Evt>>;
+
+/// Specifies the slice of the Event Stream to select when calling [`Store::stream`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum VersionSelect {
+    /// Selects all [Events] in the Event [Stream].
+    All,
+
+    /// Selects all [Events] in the Event [Stream] starting from the [Event]
+    /// with the specified [Version].
+    From(Version),
+}
```

Comment:
this looks an awful lot like `RangeBounds<Version>`

Reply:
We already discussed this here: https://github.com/get-eventually/eventually-rs/issues/49#issuecomment-927315293 👀

Reply:
in the case that you're confident these are the only two options you'll ever need, this too starts looking like an `Option<Version>`.

as in,

```rust
    /// Opens an Event Stream, effectively streaming all Domain Events
    /// of an Event Stream back in the application.
    ///
    /// If provided, the 'lower bound' defines the *earliest* included event.
    fn stream(
        &self,
        id: &Self::StreamId,
        lower_bound: Option<Version>,
    ) -> Stream<Self::StreamId, Self::Event, Self::StreamError>;
```

Reply:
This doesn't look great in terms of readability tho... between `Some(Version(2))` and `VersionSelect::From(Version(2))` the second looks a bit more descriptive to me 😅 Maybe the naming of the type can be improved tho...

Reply:
i think it would be clear in context

```rust
let lower_bound = Some(Version(2));
store.stream(id, lower_bound)
```

`VersionSelect::From(Version(2))` is closer to english, but adds additional types and complexity that a third-party developer has to understand and parse. They already know what an `Option` is, and how it behaves ;)

Reply:
I prefer the explicit name than using `Option` with an unclear API just because it's a more widespread type honestly.